### ISSUE TITLE
Create an API to manage synonyms

### ISSUE THREAD

**Issue by sdavids13** [elastic/elasticsearch#38523]
We would like the ability to manage synonyms via an API call instead of relying on putting a file down on all of the elasticsearch nodes or specifying in-line synonym in the analysis chain of a new index which isn't updatable. Ideally this would work similarly to Solr's ["Managed Resource Filters"](https://lucene.apache.org/solr/guide/6_6/managed-resources.html) which allows the ability to modify synonyms and stopwords.

**Comment by frutik** [comment]
Would be great to have a list of stopwords in a separate index too

**Comment by peterdm** [comment]
This is especially important for Elastic Cloud which has no programmatic way to upload changes to synonym files (managed manually through console/plugin-management).  

This limitation forces integration with upstream metadata management systems to happen directly in the analyzer chain (something Elastic has flagged as a bad practice since it increases the size of the cluster state.)

**Comment by schonert** [comment]
@elastic/es-search any news on this?

**Comment by Henr1k80** [comment]
> This is especially important for Elastic Cloud which has no programmatic way to upload changes to synonym files (managed manually through console/plugin-management).
As I understand it, plugins are only synced on start of the cluster, right? Meaning you have to restart the cluster to get the new files.

**Comment by Henr1k80** [comment]
Could a solution be to make a plugin with a Synonym graph token filter that reads synonyms of the cluster settings?
E.g.
```
        "filter": {
          "graph_synonyms": {
            "type": "synonym_graph_clustersettings",
            "updateable": "true",
            "lenient": "true",
            "synonyms_clustersettings": "dictionaries.english_synonyms"
          }
        }
```
and then put synonyms there:
```
PUT /_cluster/settings
{
  "persistent" : {
    "dictionaries.english_synonyms" : [
                "lol, laughing out loud",
                "universe, cosmos",
                "i-pod, i pod => ipod"
              ]
  }
}
```
The format could also just be a multi line string, if that is better.
I know that this goes against the advice of having a small cluster state, but we do not need a complete generic synonyms dictionary, only ones we explicitly create. For our e-commerce use, we only need relatively few that matches that exact catalogs content, like 200 max.
The plugin could come with all sorts of warnings, saying that you should not add a huge list and keep the cluster state small etc.

**Comment by Henr1k80** [comment]
A better solution could be to store the data in a separate index
E.g.:
```
        "filter": {
          "graph_synonyms": {
            "type": "synonym_graph_indexsource",
            "updateable": "true",
            "lenient": "true",
            "synonyms_clustersettings": "dictionaries/english_synonyms"
          }
        }
```
and then put synonyms there:
```
PUT /dictionaries/_doc/english_synonyms
{
  "synonyms": [
                "lol, laughing out loud",
                "universe, cosmos",
                "i-pod, i pod => ipod"
              ]
}
```
that also support partial updates, making your API for maintaining synonyms much easier.
The same document could also contain stopwords and keywords if needed.

**Comment by Henr1k80** [comment]
I have made a plugin with a filter that works like I mentioned in the last comment.
The index with the filter will fail to come up until the dictionary index is up and throws some exceptions in the logs, but good enough for our use™. Maybe a real elasticsearch developer can help delay the init of the index until dictionary index is up or at least make it less noisy in the logs.
Would you like me to move the code from a plugin to inside elasticsearch and make a pull request?
From 7.x branch?

**Comment by cbuescher** [comment]
> I have made a plugin with a filter that works like I mentioned in the last comment.
> The index with the filter will fail to come up until the dictionary index is up and throws some exceptions in the logs, but good enough for our use™

Thanks for trying this out, and yes, the dependency on an external index is one of the many problems to think about when thinking about loading analyzer resources from other indices. Unfortunately, today it is no straight forward way to ensure a certain ordering of index intitializations, this is one ot the things a potential solution would have to solve. I'm glad it solves your usecase though.

> Would you like me to move the code from a plugin to inside elasticsearch and make a pull request?

We always welcome pull requests, however I don't want to promise anything in terms of the time someone might need to take a look at this or if the choices you made for your plugin (e.g. storing all synonyms in one document as your example suggests) is the way we envision a longer-standing solution for this to work. Please keep that in mind before spending considerable amount of work converting what you have to a PR.

**Comment by Henr1k80** [comment]
> Thanks for trying this out, and yes, the dependency on an external index is one of the many problems to think about when thinking about loading analyzer resources from other indices.

Just to clarify, I guess it makes no difference if synonyms live a document in the index itself? Or is there some hooks that can reload the analyzer post index init?

> We always welcome pull requests, however I don't want to promise anything in terms of the time someone might need to take a look at this or if the choices you made for your plugin (e.g. storing all synonyms in one document as your example suggests) is the way we envision a longer-standing solution for this to work. Please keep that in mind before spending considerable amount of work converting what you have to a PR.

Thanks for the honesty, I will skip the PR
I was hoping the pull request would "just" be about naming, documentation, test coverage etc. 😅

**Comment by nhsome** [comment]
@Henr1k80, where can I see the source code of your plugin. Potentially this will be a good solution for me.

**Comment by Danouchka** [comment]
Any news regarding this ER please ?

**Comment by berkayalcin** [comment]
Any Updates ?

**Comment by sdavids13** [comment]
FYI here is what AWS OpenSearch does for inspiration: https://docs.aws.amazon.com/opensearch-service/latest/developerguide/custom-packages.html#custom-packages-updating

**Comment by Henr1k80** [comment]
> FYI here is what AWS OpenSearch does for inspiration: https://docs.aws.amazon.com/opensearch-service/latest/developerguide/custom-packages.html#custom-packages-updating

Last time I checked, any updates to the plugins required a restart of the deployment to sync the plugin files to the nodes.
Not a viable solution

**Comment by Danouchka** [comment]
Any news please ?

**Comment by SiwyDym** [comment]
Do you have some news about API? :)

**Comment by Henr1k80** [comment]
It is now in beta in 8.10 🥳
https://www.elastic.co/guide/en/elasticsearch/reference/8.10/synonyms-apis.html

**Comment by mayya-sharipova** [comment]
Indeed,  it is implemented and tech preview is available in 8.10. Closing the issue
